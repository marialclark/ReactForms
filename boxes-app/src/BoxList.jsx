import { useState } from 'react';
import Box from './Box.jsx';
import NewBoxForm from './NewBoxForm.jsx';
import './styles/BoxList.css';

const BoxList = () => {
  const [boxes, setBoxes] = useState([]);

  // Function for adding new box
  const addBox = (newBox) => {
    setBoxes((boxes) => [...boxes, newBox]);
  };

  // Function for removing a box (by ID)
  const removeBox = (id) => {
    setBoxes((boxes) => boxes.filter((box) => box.id !== id));
  };

  return (
    <div className="BoxList">
      <NewBoxForm addBox={addBox} />
      <div className="BoxList-container">
        {boxes.map((box) => (
          <Box 
            key={box.id}
            id={box.id}
            width={box.width}
            height={box.height}
            backgroundColor={box.backgroundColor}
            removeBox={removeBox}
          />
        ))}
      </div>
    </div>
  );
}

export default BoxList;