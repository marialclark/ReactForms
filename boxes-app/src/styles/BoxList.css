.BoxList {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 20px;
}

.BoxList-container {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: flex-start;
  gap: 20px;
}
