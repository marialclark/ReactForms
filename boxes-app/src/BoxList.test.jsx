import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

// Helper function to add a box
function addBox(height = '100', width = '150', color = 'blue') {
  fireEvent.change(screen.getByLabelText('Height:'), { target: { value: height } });
  fireEvent.change(screen.getByLabelText('Width:'), { target: { value: width } });
  fireEvent.change(screen.getByLabelText('Box Color:'), { target: { value: color } });
  fireEvent.click(screen.getByText('Add Box'));
}

// Smoke Test
it('renders without crashing', () => {
  render(<BoxList />);
});

// Snapshot Test
it('matches snapshot', () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

// Critical Business Logic: Adding a Box
it('adds a new box to the list', () => {
  render(<BoxList />);

  expect(screen.queryByText('X')).not.toBeInTheDocument();

  addBox();

  const removeButton = screen.getByText('X');
  expect(removeButton).toBeInTheDocument();
  expect(removeButton.previousSibling).toHaveStyle(`
    width: 150px;
    height: 100px;
    background-color: blue;
  `);
});

// Critical Business Logic: Removing a Box
it('removes a box from the list', () => {
  render(<BoxList />);

  addBox();

  const removeButton = screen.getByText('X');
  fireEvent.click(removeButton);

  expect(removeButton).not.toBeInTheDocument();
});
