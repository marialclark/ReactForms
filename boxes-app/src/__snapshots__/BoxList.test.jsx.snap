// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form
      class="NewBoxForm"
    >
      <div
        class="NewBoxForm-field"
      >
        <label
          for="width"
        >
          Width:
        </label>
        <input
          id="width"
          name="width"
          type="text"
          value=""
        />
      </div>
      <div
        class="NewBoxForm-field"
      >
        <label
          for="height"
        >
          Height:
        </label>
        <input
          id="height"
          name="height"
          type="text"
          value=""
        />
      </div>
      <div
        class="NewBoxForm-field"
      >
        <label
          for="backgroundColor"
        >
          Box Color:
        </label>
        <input
          id="backgroundColor"
          name="backgroundColor"
          type="text"
          value=""
        />
      </div>
      <div
        class="NewBoxForm-button"
      >
        <button>
          Add Box
        </button>
      </div>
    </form>
    <div
      class="BoxList-container"
    />
  </div>
</DocumentFragment>
`;
